checkbox-ng/plainbox/impl/depmgr.py

```diff
@@ -320,93 +336,217 @@ def __init__(self, job_list):
         self._job_list = job_list
         # Build a map of jobs (by id)
         self._job_map = self._get_job_map(job_list)
-        # Job colors, maps from job.id to COLOR_xxx
-        self._job_color_map = {job.id: self.COLOR_WHITE for job in job_list}
-        # The computed solution, made out of job instances. This is not
-        # necessarily the only solution but the algorithm computes the same
-        # value each time, given the same input.
-        self._solution = []
+
+        # Solution for all the dependencies that may pull in new jobs
+        self._pull_solution = []
+        # The computed solution for all dependencies
+        self._order_solution = []
+
+    def _clear_state_map(self):
+        self._job_state_map = {
+            job.id: State.NOT_VISITED for job in self._job_list
+        }
 
     def _solve(self, visit_list=None):
         """
         Internal method of DependencySolver.
 
         Solves the dependency graph and returns the solution.
+        The algorithm runs in two passes:
+        1. Pulls all the dependencies of the initial jobs and builds a solution
+        2. Orders the jobs in the solution according to their dependencies
+
+        The two passes are required to treat the dependencies that pull in
+        new jobs differently from the dependencies that only change the order
 
         Calls _visit() on each of the initial nodes/jobs
         """
         # Visit the visit list
         logger.debug(_("Starting solve"))
         logger.debug(_("Solver job list: %r"), self._job_list)
         logger.debug(_("Solver visit list: %r"), visit_list)
+
+        # If no visit list is specified, use the full job list
         if visit_list is None:
-            visit_list = self._job_list
-        for job in visit_list:
-            self._visit(job=job, visit_list=visit_list)
+            self._visit_list = self._job_list
+        else:
+            self._visit_list = visit_list
+
+        # Solve first for pulling dependencies
+        self._clear_state_map()
+        for job in self._visit_list:
+            self._visit(job, pull=True)
+
+        # Create a map of pulled jobs
+        self._pulled_map = self._get_job_map(self._pull_solution)
+        # Add the before dependencies for the jobs in the map
+        for job in self._pull_solution:
+            job.controller.add_before_deps(
+                job, self._pulled_map, self._job_map
+            )
+
+        # Solve again for order dependencies, using the pulled jobs as the
+        # new visit list
+        self._clear_state_map()
+        for job in self._pull_solution:
+            self._visit(job)
+
+        # Perform a sanity check to ensure that no jobs have been added or
+        # removed from the solution.
+        if len(self._order_solution) != len(self._pull_solution):
+            raise ValueError(
```

Comment:
I don't know how this can happen, but I know the output would be better if it contained the diff of `self._order_solution` and `self._pull_solution`, because otherwise debugging will be difficult.